NXOS: Manually Configurable Route-Target

richardstrnad opened the pull request:
##### SUMMARY
Provides the option to manually add route-targets under an address family on Cisco Nexus (NXOS) Devices.

Fixes #41397

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
nxos_vrf_af

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

richardstrnad committed:
Renamed auto evpn test

richardstrnad committed:
Made sure that the current module is idempotent with additional tests

richardstrnad committed:
Added tests for route-target import function

richardstrnad committed:
Added tests for route-target export function

richardstrnad committed:
Added tests for route-target both function

richardstrnad committed:
PEP8 syntax fix

richardstrnad committed:
Added route-target import & export function

richardstrnad committed:
Added required 'version_added: "2.8"' in the documentation

richardstrnad committed:
Updated documentation of new route-target options

mikewiebe commented on lib/ansible/modules/network/nxos/nxos_vrf_af.py:
```diff
@@ -61,6 +83,46 @@
+    afi: ipv4
+    route_target_both:
+      - rt: '65000:1000'
+        state: present
```
How does the per route_target `state` work with the overall `state` property?  For example what should I expect in the following case:

```yaml
- nxos_vrf_af:
    state: absent
    vrf: ntc
    afi: ipv4
    route_target_export:
      - rt: '65000:1000'
        state: present
```

richardstrnad replied:
Good point. 

Route-targets exist under the address-family, therefore they can't exists without the address-family. In case of an `absent` address-family, the route-target part (same with `route_target_both_auto_evpn`) is ignored.

mikewiebe replied:
@richardstrnad I suspected that was the desired workflow so thanks for confirming.  Could you add a note in the doc header to clarify?

trishnaguha commented:
cc @cidrblock

richardstrnad committed:
Added a test to make sure that in case of `state=absent` on the vrf
level the route-target options are ignored.

richardstrnad committed:
Specified that the route-target options are ignored in case of
`state=absent'.

richardstrnad committed:
Updated the doc to the correct format (using 'C()')

trishnaguha reviewed (commented):
Instead of having three different params, we can have `route_target: [{routes: '...', direction: '{import|export|both}', state: '...'}]`?

richardstrnad commented:
> Instead of having three different params, we can have `route_target: [{routes: '...', direction: '{import|export|both}', state: '...'}]`?

I have no preference, but maybe that makes more sense :-)
Should i change the PR?

trishnaguha commented:
@richardstrnad That would be great. Thanks!

clammers commented:
Could this be enhanced with the route-target import/export evpn command?

richardstrnad committed:
Changed the VRF Route Target Syntax

Instead of using three different params (route_target_import,
route_target_export, route_target_both) the module uses now only one
param (route_targets) and the direction is specified for each of the
route targets.

Example:
route_targets:
[{rt: '...', direction: '{import|export|both}', state: '...'}]

richardstrnad committed:
Updated Description and Examples to reflect new params

richardstrnad committed:
Updated "version_added"

richardstrnad committed:
pep8 fixes

richardstrnad committed:
If rt['direction'] is not definied, we assume default 'both' and run
the same routine

richardstrnad committed:
Added test with default direction for route-targets

richardstrnad commented:
> Could this be enhanced with the route-target import/export evpn command?

I think it make sense to create a dedicated issue/pr for that.

richardstrnad commented:
> @richardstrnad That would be great. Thanks!

Finally had time, PR updated :-)

richardstrnad commented:
Anything more required by me?

NilashishC commented:
recheck

NilashishC commented:
Hi @richardstrnad. Thank you for this PR. Apologize for the delay on this one. As it appears, Shippable is failing with some sanity tests issues for this module. Could you have a look at it?

richardstrnad committed:
Documentation fixes
